File: src/main/kotlin/top/bearingwall/asya/service/RoundService.kt
package top.bearingwall.asya.service

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import top.bearingwall.asya.audit.Auditable
import top.bearingwall.asya.dto.RoundPublishRequest
import top.bearingwall.asya.dto.RoundResponse
import top.bearingwall.asya.dto.RoundSetNextRequest
import top.bearingwall.asya.dto.RoundSetCurrentRequest
import top.bearingwall.asya.dto.RoundUpdateRequest
import top.bearingwall.asya.model.AuditActionType
import top.bearingwall.asya.model.Round
import top.bearingwall.asya.model.RoundStatus
import top.bearingwall.asya.repository.ConferenceRepository
import top.bearingwall.asya.repository.RoundRepository
import java.time.Duration
import java.time.LocalDateTime
import java.util.UUID
import kotlin.math.max

@Service
class RoundService(
    private val roundRepository: RoundRepository,
    private val conferenceRepository: ConferenceRepository
) {
    private val log = LoggerFactory.getLogger(RoundService::class.java)

    @Transactional
    @Auditable(type = AuditActionType.ROUND_PUBLISH, content = "发布回合")
    fun publishRound(request: RoundPublishRequest, conferenceUuid: UUID): RoundResponse {
        require(request.name.isNotBlank()) { "Round name cannot be blank" }
        require(request.durationSeconds > 0) { "durationSeconds must be greater than 0" }

        val conference = conferenceRepository.findById(conferenceUuid).orElseThrow {
            IllegalArgumentException("Conference not found: $conferenceUuid")
        }

        val nextRound = parseNextRound(request.nextRoundId, conferenceUuid)
        val now = LocalDateTime.now()

        roundRepository.updateIsCurrentToFalseByConferenceUuid(conferenceUuid)

        val round = Round(
            conference = conference,
            name = request.name.trim(),
            durationSeconds = request.durationSeconds,
            remainingSeconds = request.durationSeconds,
            status = request.initialStatus,
            isCurrent = true,
            endAt = if (request.initialStatus == RoundStatus.RUNNING) now.plusSeconds(request.durationSeconds) else null,
            updatedAt = now,
            nextRound = nextRound
        )

        val saved = roundRepository.save(round)
        return saved.toResponse(now)
    }

    @Transactional
    @Auditable(type = AuditActionType.ROUND_SET_NEXT, content = "设置回合下一跳")
    fun setNextRound(roundUuid: UUID, request: RoundSetNextRequest, conferenceUuid: UUID): RoundResponse {
        val round = roundRepository.findByUuidAndConferenceUuid(roundUuid, conferenceUuid)
            ?: throw IllegalArgumentException("Round not found: $roundUuid")

        round.nextRound = parseNextRound(request.nextRoundId, conferenceUuid)
        round.updatedAt = LocalDateTime.now()

        return roundRepository.save(round).toResponse(LocalDateTime.now())
    }

    @Transactional
    @Auditable(type = AuditActionType.ROUND_UPDATE, content = "修改回合")
    fun updateRound(roundUuid: UUID, request: RoundUpdateRequest, conferenceUuid: UUID): RoundResponse {
        advanceIfExpired(conferenceUuid)

        val round = roundRepository.findByUuidAndConferenceUuid(roundUuid, conferenceUuid)
            ?: throw IllegalArgumentException("Round not found: $roundUuid")

        return updateRound(round, request)
    }

    @Transactional
    @Auditable(type = AuditActionType.ROUND_SET_CURRENT, content = "切换当前回合")
    fun setCurrentRound(request: RoundSetCurrentRequest, conferenceUuid: UUID): RoundResponse {
        advanceIfExpired(conferenceUuid)

        val targetRoundUuid = runCatching { UUID.fromString(request.roundId) }
            .getOrElse { throw IllegalArgumentException("Invalid roundId") }
        val now = LocalDateTime.now()

        val target = roundRepository.findByUuidAndConferenceUuid(targetRoundUuid, conferenceUuid)
            ?: throw IllegalArgumentException("Round not found: $targetRoundUuid")

        val current = roundRepository.findCurrentForUpdate(conferenceUuid)
        if (current?.uuid == target.uuid) {
            return target.toResponse(now)
        }

        if (current != null) {
            current.isCurrent = false
            if (current.status == RoundStatus.RUNNING) {
                current.remainingSeconds = remainingSeconds(current, now)
                current.status = RoundStatus.PAUSED
                current.endAt = null
            }
            current.updatedAt = now
            roundRepository.save(current)
        }

        target.isCurrent = true
        if (target.status == RoundStatus.RUNNING) {
            target.endAt = now.plusSeconds(target.remainingSeconds)
        } else {
            target.endAt = null
        }
        target.updatedAt = now

        return roundRepository.save(target).toResponse(now)
    }

    @Transactional
    @Auditable(type = AuditActionType.ROUND_PAUSE, content = "暂停回合")
    fun pauseRound(roundUuid: UUID, conferenceUuid: UUID): RoundResponse {
        advanceIfExpired(conferenceUuid)

        val current = roundRepository.findCurrentForUpdate(conferenceUuid)
            ?: throw IllegalStateException("No current round")
        require(current.uuid == roundUuid) { "Only current round can be paused" }
        require(current.status == RoundStatus.RUNNING) { "Round is not running" }

        val now = LocalDateTime.now()
        current.remainingSeconds = remainingSeconds(current, now)
        current.status = RoundStatus.PAUSED
        current.endAt = null
        current.updatedAt = now

        return roundRepository.save(current).toResponse(now)
    }

    @Transactional
    @Auditable(type = AuditActionType.ROUND_RESUME, content = "恢复回合")
    fun resumeRound(roundUuid: UUID, conferenceUuid: UUID): RoundResponse {
        advanceIfExpired(conferenceUuid)

        val current = roundRepository.findCurrentForUpdate(conferenceUuid)
            ?: throw IllegalStateException("No current round")
        require(current.uuid == roundUuid) { "Only current round can be resumed" }
        require(current.status == RoundStatus.PAUSED) { "Round is not paused" }
        require(current.remainingSeconds > 0) { "Round has no remaining time" }

        val now = LocalDateTime.now()
        current.status = RoundStatus.RUNNING
        current.endAt = now.plusSeconds(current.remainingSeconds)
        current.updatedAt = now

        return roundRepository.save(current).toResponse(now)
    }

    @Transactional(readOnly = true)
    fun listRounds(conferenceUuid: UUID): List<RoundResponse> {
        val now = LocalDateTime.now()
        return roundRepository.findAllByConferenceUuidOrderByUpdatedAtDesc(conferenceUuid).map { it.toResponse(now) }
    }

    @Transactional
    fun getRoundDetail(roundUuid: UUID, conferenceUuid: UUID): RoundResponse {
        advanceIfExpired(conferenceUuid)
        val round = roundRepository.findByUuidAndConferenceUuid(roundUuid, conferenceUuid)
            ?: throw IllegalArgumentException("Round not found: $roundUuid")
        return round.toResponse(LocalDateTime.now())
    }

    @Transactional
    fun getCurrentRound(conferenceUuid: UUID): RoundResponse? {
        advanceIfExpired(conferenceUuid)
        val now = LocalDateTime.now()
        return roundRepository.findFirstByConferenceUuidAndIsCurrentTrue(conferenceUuid)?.toResponse(now)
    }

    @Transactional
    @Auditable(type = AuditActionType.ROUND_AUTO_ADVANCE, content = "回合自动推进")
    fun advanceIfExpired(conferenceUuid: UUID): RoundResponse? {
        val now = LocalDateTime.now()
        val current = roundRepository.findCurrentForUpdate(conferenceUuid) ?: return null

        if (current.status != RoundStatus.RUNNING || current.endAt == null || current.endAt!!.isAfter(now)) {
            return current.toResponse(now)
        }

        current.isCurrent = false
        current.status = RoundStatus.PAUSED
        current.remainingSeconds = 0
        current.endAt = null
        current.updatedAt = now
        roundRepository.save(current)

        val nextRoundId = current.nextRound?.uuid
        if (nextRoundId == null) {
            return null
        }

        val nextRound = roundRepository.findByUuidAndConferenceUuid(nextRoundId, conferenceUuid)
            ?: throw IllegalStateException("Next round not found in same conference: $nextRoundId")

        roundRepository.updateIsCurrentToFalseByConferenceUuid(conferenceUuid)

        nextRound.isCurrent = true
        nextRound.status = RoundStatus.RUNNING
        nextRound.remainingSeconds = nextRound.durationSeconds
        nextRound.endAt = now.plusSeconds(nextRound.durationSeconds)
        nextRound.updatedAt = now

        return roundRepository.save(nextRound).toResponse(now)
    }

    @Transactional
    fun advanceExpiredRounds() {
        val now = LocalDateTime.now()
        val conferenceIds = roundRepository.findConferenceIdsWithExpiredCurrentRound(now)
        conferenceIds.forEach { conferenceId ->
            runCatching { advanceIfExpired(conferenceId) }
                .onFailure { ex ->
                    log.error("Failed to auto advance round for conference {}", conferenceId, ex)
                }
        }
    }

    private fun parseNextRound(nextRoundId: String?, conferenceUuid: UUID): Round? {
        if (nextRoundId.isNullOrBlank()) {
            return null
        }
        val nextUuid = runCatching { UUID.fromString(nextRoundId) }
            .getOrElse { throw IllegalArgumentException("Invalid nextRoundId") }

        return roundRepository.findByUuidAndConferenceUuid(nextUuid, conferenceUuid)
            ?: throw IllegalArgumentException("nextRoundId not found in current conference")
    }

    private fun updateRound(round: Round, request: RoundUpdateRequest): RoundResponse {
        require(request.name.isNotBlank()) { "Round name cannot be blank" }
        require(request.durationSeconds > 0) { "durationSeconds must be greater than 0" }

        val now = LocalDateTime.now()
        val currentRemaining = remainingSeconds(round, now)
        val elapsedSeconds = max(0, round.durationSeconds - currentRemaining)
        val newRemainingSeconds = max(0, request.durationSeconds - elapsedSeconds)

        round.name = request.name.trim()
        round.durationSeconds = request.durationSeconds
        round.remainingSeconds = newRemainingSeconds
        round.updatedAt = now

        if (round.status == RoundStatus.RUNNING) {
            round.endAt = now.plusSeconds(newRemainingSeconds)
        }

        if (round.status == RoundStatus.PAUSED) {
            round.endAt = null
        }

        return roundRepository.save(round).toResponse(now)
    }

    private fun remainingSeconds(round: Round, now: LocalDateTime): Long {
        return when (round.status) {
            RoundStatus.PAUSED -> round.remainingSeconds
            RoundStatus.RUNNING -> {
                val endAt = round.endAt ?: return round.remainingSeconds
                max(0, Duration.between(now, endAt).seconds)
            }
        }
    }

    private fun Round.toResponse(now: LocalDateTime): RoundResponse {
        val conferenceId = this.conference?.uuid?.toString()
            ?: throw IllegalStateException("Round conference is missing")
        val roundId = this.uuid?.toString() ?: throw IllegalStateException("Round uuid is missing")

        return RoundResponse(
            roundId = roundId,
            conferenceId = conferenceId,
            name = this.name,
            durationSeconds = this.durationSeconds,
            remainingSeconds = remainingSeconds(this, now),
            status = this.status,
            isCurrent = this.isCurrent,
            nextRoundId = this.nextRound?.uuid?.toString(),
            endAt = this.endAt,
            serverTime = now
        )
    }
}

File: backend-dotnet/src/AsyaMun.Api/Services/RoundService.cs
using System.Data;
using AsyaMun.Api.Data;
using AsyaMun.Api.Dtos;
using AsyaMun.Api.Errors;
using AsyaMun.Api.Models;
using Microsoft.EntityFrameworkCore;

namespace AsyaMun.Api.Services;

public class RoundService
{
    private readonly AppDbContext _db;
    private readonly ILogger<RoundService> _logger;

    public RoundService(AppDbContext db, ILogger<RoundService> logger)
    {
        _db = db;
        _logger = logger;
    }

    public async Task<RoundResponse> PublishRoundAsync(RoundPublishRequest request, Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            if (string.IsNullOrWhiteSpace(request.Name))
            {
                throw AsyaBusinessException.ParamError("Round name cannot be blank");
            }

            if (request.DurationSeconds <= 0)
            {
                throw AsyaBusinessException.ParamError("durationSeconds must be greater than 0");
            }

            var conferenceExists = await _db.Conferences.AsNoTracking().AnyAsync(c => c.Uuid == conferenceUuid);
            if (!conferenceExists)
            {
                throw AsyaBusinessException.ParamError($"Conference not found: {conferenceUuid}");
            }

            var nextRoundId = await ParseNextRoundAsync(request.NextRoundId, conferenceUuid);
            var now = NowUtc();

            await ClearCurrentRoundsAsync(conferenceUuid);

            var round = new Round
            {
                ConferenceId = conferenceUuid,
                Name = request.Name.Trim(),
                DurationSeconds = request.DurationSeconds,
                RemainingSeconds = request.DurationSeconds,
                Status = request.InitialStatus,
                IsCurrent = true,
                EndAt = request.InitialStatus == RoundStatus.RUNNING
                    ? now.AddSeconds(request.DurationSeconds)
                    : null,
                UpdatedAt = now,
                NextRoundId = nextRoundId
            };

            _db.Rounds.Add(round);
            await _db.SaveChangesAsync();

            return ToResponse(round, now);
        });
    }

    public async Task<RoundResponse> SetNextRoundAsync(Guid roundUuid, RoundSetNextRequest request, Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            var round = await FindByUuidAndConferenceAsync(roundUuid, conferenceUuid)
                ?? throw AsyaBusinessException.ParamError($"Round not found: {roundUuid}");

            round.NextRoundId = await ParseNextRoundAsync(request.NextRoundId, conferenceUuid);
            round.UpdatedAt = NowUtc();
            await _db.SaveChangesAsync();

            return ToResponse(round, NowUtc());
        });
    }

    public async Task<RoundResponse> UpdateRoundAsync(Guid roundUuid, RoundUpdateRequest request, Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            await AdvanceIfExpiredAsync(conferenceUuid);

            var round = await FindByUuidAndConferenceAsync(roundUuid, conferenceUuid)
                ?? throw AsyaBusinessException.ParamError($"Round not found: {roundUuid}");

            return await UpdateRoundCoreAsync(round, request);
        });
    }

    public async Task<RoundResponse> SetCurrentRoundAsync(RoundSetCurrentRequest request, Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            await AdvanceIfExpiredAsync(conferenceUuid);

            if (!Guid.TryParse(request.RoundId, out var targetRoundUuid))
            {
                throw AsyaBusinessException.ParamError("Invalid roundId");
            }

            var now = NowUtc();

            var target = await FindByUuidAndConferenceAsync(targetRoundUuid, conferenceUuid)
                ?? throw AsyaBusinessException.ParamError($"Round not found: {targetRoundUuid}");

            var current = await FindCurrentAsync(conferenceUuid);
            if (current?.Uuid == target.Uuid)
            {
                return ToResponse(target, now);
            }

            if (current != null)
            {
                current.IsCurrent = false;
                if (current.Status == RoundStatus.RUNNING)
                {
                    current.RemainingSeconds = RemainingSeconds(current, now);
                    current.Status = RoundStatus.PAUSED;
                    current.EndAt = null;
                }

                current.UpdatedAt = now;
            }

            target.IsCurrent = true;
            target.EndAt = target.Status == RoundStatus.RUNNING
                ? now.AddSeconds(target.RemainingSeconds)
                : null;
            target.UpdatedAt = now;

            await _db.SaveChangesAsync();

            return ToResponse(target, now);
        });
    }

    public async Task<RoundResponse> SetRoundRemainingAsync(Guid roundUuid, RoundSetRemainingRequest request, Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            if (request.RemainingSeconds < 0)
            {
                throw AsyaBusinessException.ParamError("remainingSeconds must be greater than or equal to 0");
            }

            await AdvanceIfExpiredAsync(conferenceUuid);

            var round = await FindByUuidAndConferenceAsync(roundUuid, conferenceUuid)
                ?? throw AsyaBusinessException.ParamError($"Round not found: {roundUuid}");

            var now = NowUtc();
            round.RemainingSeconds = request.RemainingSeconds;
            round.EndAt = round.Status == RoundStatus.RUNNING
                ? now.AddSeconds(request.RemainingSeconds)
                : null;
            round.UpdatedAt = now;
            await _db.SaveChangesAsync();

            return ToResponse(round, now);
        });
    }

    public async Task<RoundResponse> PauseRoundAsync(Guid roundUuid, Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            await AdvanceIfExpiredAsync(conferenceUuid);

            var current = await FindCurrentAsync(conferenceUuid)
                ?? throw new InvalidOperationException("No current round");
            if (current.Uuid != roundUuid)
            {
                throw AsyaBusinessException.ParamError("Only current round can be paused");
            }

            if (current.Status != RoundStatus.RUNNING)
            {
                throw AsyaBusinessException.ParamError("Round is not running");
            }

            var now = NowUtc();
            current.RemainingSeconds = RemainingSeconds(current, now);
            current.Status = RoundStatus.PAUSED;
            current.EndAt = null;
            current.UpdatedAt = now;
            await _db.SaveChangesAsync();

            return ToResponse(current, now);
        });
    }

    public async Task<RoundResponse> ResumeRoundAsync(Guid roundUuid, Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            await AdvanceIfExpiredAsync(conferenceUuid);

            var current = await FindCurrentAsync(conferenceUuid)
                ?? throw new InvalidOperationException("No current round");
            if (current.Uuid != roundUuid)
            {
                throw AsyaBusinessException.ParamError("Only current round can be resumed");
            }

            if (current.Status != RoundStatus.PAUSED)
            {
                throw AsyaBusinessException.ParamError("Round is not paused");
            }

            var now = NowUtc();
            if (current.RemainingSeconds <= 0)
            {
                current.RemainingSeconds = current.DurationSeconds;
            }

            current.Status = RoundStatus.RUNNING;
            current.EndAt = now.AddSeconds(current.RemainingSeconds);
            current.UpdatedAt = now;
            await _db.SaveChangesAsync();

            return ToResponse(current, now);
        });
    }

    public async Task<List<RoundResponse>> ListRoundsAsync(Guid conferenceUuid)
    {
        var now = NowUtc();
        var rounds = await _db.Rounds.AsNoTracking()
            .Where(r => r.ConferenceId == conferenceUuid)
            .OrderByDescending(r => r.UpdatedAt)
            .ToListAsync();

        return rounds.Select(r => ToResponse(r, now)).ToList();
    }

    public async Task<RoundResponse> GetRoundDetailAsync(Guid roundUuid, Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            await AdvanceIfExpiredAsync(conferenceUuid);

            var round = await FindByUuidAndConferenceAsync(roundUuid, conferenceUuid)
                ?? throw AsyaBusinessException.ParamError($"Round not found: {roundUuid}");

            return ToResponse(round, NowUtc());
        });
    }

    public async Task<RoundResponse?> GetCurrentRoundAsync(Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            await AdvanceIfExpiredAsync(conferenceUuid);

            var now = NowUtc();
            var round = await _db.Rounds.AsNoTracking()
                .FirstOrDefaultAsync(r => r.ConferenceId == conferenceUuid && r.IsCurrent);

            return round == null ? null : ToResponse(round, now);
        });
    }

    public async Task<RoundResponse?> AdvanceIfExpiredAsync(Guid conferenceUuid)
    {
        return await ExecuteAsync(async () =>
        {
            var now = NowUtc();
            var current = await FindCurrentAsync(conferenceUuid);
            if (current == null)
            {
                return null;
            }

            if (current.Status != RoundStatus.RUNNING || current.EndAt == null || current.EndAt.Value > now)
            {
                return ToResponse(current, now);
            }

            current.IsCurrent = false;
            current.Status = RoundStatus.PAUSED;
            current.RemainingSeconds = 0;
            current.EndAt = null;
            current.UpdatedAt = now;
            await _db.SaveChangesAsync();

            var nextRoundId = current.NextRoundId;
            if (nextRoundId == null)
            {
                return null;
            }

            var nextRound = await FindByUuidAndConferenceAsync(nextRoundId.Value, conferenceUuid)
                ?? throw new InvalidOperationException($"Next round not found in same conference: {nextRoundId}");

            await ClearCurrentRoundsAsync(conferenceUuid);

            nextRound.IsCurrent = true;
            nextRound.Status = RoundStatus.RUNNING;
            nextRound.RemainingSeconds = nextRound.DurationSeconds;
            nextRound.EndAt = now.AddSeconds(nextRound.DurationSeconds);
            nextRound.UpdatedAt = now;
            await _db.SaveChangesAsync();

            return ToResponse(nextRound, now);
        });
    }

    public async Task AdvanceExpiredRoundsAsync()
    {
        var now = NowUtc();
        var conferenceIds = await _db.Rounds.AsNoTracking()
            .Where(r => r.IsCurrent
                && r.Status == RoundStatus.RUNNING
                && r.EndAt != null
                && r.EndAt <= now)
            .Select(r => r.ConferenceId)
            .Distinct()
            .ToListAsync();

        foreach (var conferenceId in conferenceIds)
        {
            try
            {
                await AdvanceIfExpiredAsync(conferenceId);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to auto advance round for conference {ConferenceId}", conferenceId);
            }
        }
    }

    private async Task<Guid?> ParseNextRoundAsync(string? nextRoundId, Guid conferenceUuid)
    {
        if (string.IsNullOrWhiteSpace(nextRoundId))
        {
            return null;
        }

        if (!Guid.TryParse(nextRoundId, out var nextUuid))
        {
            throw AsyaBusinessException.ParamError("Invalid nextRoundId");
        }

        var exists = await _db.Rounds.AsNoTracking()
            .AnyAsync(r => r.Uuid == nextUuid && r.ConferenceId == conferenceUuid);
        if (!exists)
        {
            throw AsyaBusinessException.ParamError("nextRoundId not found in current conference");
        }

        return nextUuid;
    }

    private async Task<RoundResponse> UpdateRoundCoreAsync(Round round, RoundUpdateRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Name))
        {
            throw AsyaBusinessException.ParamError("Round name cannot be blank");
        }

        if (request.DurationSeconds <= 0)
        {
            throw AsyaBusinessException.ParamError("durationSeconds must be greater than 0");
        }

        var now = NowUtc();
        var currentRemaining = RemainingSeconds(round, now);
        var elapsedSeconds = Math.Max(0, round.DurationSeconds - currentRemaining);
        var newRemainingSeconds = Math.Max(0, request.DurationSeconds - elapsedSeconds);

        round.Name = request.Name.Trim();
        round.DurationSeconds = request.DurationSeconds;
        round.RemainingSeconds = newRemainingSeconds;
        round.UpdatedAt = now;

        if (round.Status == RoundStatus.RUNNING)
        {
            round.EndAt = now.AddSeconds(newRemainingSeconds);
        }

        if (round.Status == RoundStatus.PAUSED)
        {
            round.EndAt = null;
        }

        await _db.SaveChangesAsync();

        return ToResponse(round, now);
    }

    private long RemainingSeconds(Round round, DateTime now)
    {
        if (round.Status == RoundStatus.PAUSED)
        {
            return round.RemainingSeconds;
        }

        if (round.EndAt == null)
        {
            return round.RemainingSeconds;
        }

        var seconds = (long)((round.EndAt.Value - now).TotalSeconds);
        return Math.Max(0, seconds);
    }

    private RoundResponse ToResponse(Round round, DateTime now)
    {
        return new RoundResponse(
            round.Uuid.ToString("D"),
            round.ConferenceId.ToString("D"),
            round.Name,
            round.DurationSeconds,
            RemainingSeconds(round, now),
            round.Status,
            round.IsCurrent,
            round.NextRoundId?.ToString("D"),
            round.EndAt,
            now);
    }

    private async Task<Round?> FindByUuidAndConferenceAsync(Guid roundUuid, Guid conferenceUuid)
    {
        return await _db.Rounds.FirstOrDefaultAsync(r => r.Uuid == roundUuid && r.ConferenceId == conferenceUuid);
    }

    private async Task<Round?> FindCurrentAsync(Guid conferenceUuid)
    {
        return await _db.Rounds.FirstOrDefaultAsync(r => r.ConferenceId == conferenceUuid && r.IsCurrent);
    }

    private Task<int> ClearCurrentRoundsAsync(Guid conferenceUuid)
    {
        return _db.Rounds
            .Where(r => r.ConferenceId == conferenceUuid && r.IsCurrent)
            .ExecuteUpdateAsync(s => s.SetProperty(r => r.IsCurrent, false));
    }

    private async Task<T> ExecuteAsync<T>(Func<Task<T>> action)
    {
        if (_db.Database.CurrentTransaction != null)
        {
            return await action();
        }

        await using var transaction = await _db.Database.BeginTransactionAsync(IsolationLevel.Serializable);
        var result = await action();
        await transaction.CommitAsync();
        return result;
    }

    private static DateTime NowUtc()
    {
        return DateTime.SpecifyKind(DateTime.UtcNow, DateTimeKind.Unspecified);
    }
}